### DISCUSSION TITLE
Can't build lotus, [too many errors]

### DISCUSSION THREAD

**Discussion by PNetrox** [filecoin-project/lotus#11264] [Build and Sync] [answered]
OS: Ubuntu server 20.04
Path: filecoin:~/home/filecoin/miner/lotus$

When using "make clean calibnet":

`/usr/local/go/src/math/erf.go:274:6: too many errors
make: *** [Makefile:89: lotus] Error 1`

Here whats above:

`[...]go: downloading github.com/filecoin-project/go-bitfield v0.2.4
go: downloading github.com/filecoin-project/specs-actors/v8 v8.0.1
go: downloading github.com/hashicorp/golang-lru/v2 v2.0.2
go: downloading github.com/ipfs/go-ipld-format v0.5.0
go: downloading github.com/ipfs/boxo v0.10.1
go: downloading github.com/ipld/go-car v0.6.1
go: downloading github.com/docker/go-units v0.5.0
go: downloading github.com/BurntSushi/toml v1.2.1
go: downloading github.com/dgraph-io/badger/v2 v2.2007.4
go: downloading github.com/ipfs/go-ds-badger2 v0.1.3
go: downloading github.com/ipfs/go-ds-leveldb v0.5.0
go: downloading github.com/ipfs/go-ds-measure v0.2.0
go: downloading github.com/ipfs/go-fs-lock v0.0.7
go: downloading github.com/multiformats/go-base32 v0.1.0
go: downloading github.com/syndtr/goleveldb v1.0.1-0.20210819022825-2ae1ddf74ef7
go: downloading github.com/hashicorp/golang-lru v0.6.0
go: downloading github.com/filecoin-project/go-data-transfer/v2 v2.0.0-rc7
go: downloading github.com/filecoin-project/go-fil-markets v1.28.3
go: downloading github.com/filecoin-project/go-statestore v0.2.0
go: downloading github.com/gbrlsnchs/jwt/v3 v3.0.1
go: downloading github.com/ipfs/go-graphsync v0.14.6
go: downloading github.com/kelseyhightower/envconfig v1.4.0
go: downloading github.com/filecoin-project/pubsub v1.0.0
go: downloading github.com/ipfs/go-block-format v0.1.2
go: downloading github.com/ipld/go-car/v2 v2.10.1
go: downloading github.com/multiformats/go-multihash v0.2.3
go: downloading github.com/puzpuzpuz/xsync/v2 v2.4.0
go: downloading go.uber.org/atomic v1.11.0
go: downloading github.com/detailyang/go-fallocate v0.0.0-20180908115635-432fa640bd2e
go: downloading github.com/filecoin-project/go-commp-utils v0.1.3
go: downloading github.com/filecoin-project/go-fil-commcid v0.1.0
go: downloading golang.org/x/sync v0.2.0
go: downloading github.com/minio/blake2b-simd v0.0.0-20160723061019-3f5f724cb5b1
go: downloading github.com/icza/backscanner v0.0.0-20210726202459-ac2ffc679f94
go: downloading github.com/raulk/clock v1.1.0
go: downloading golang.org/x/sys v0.9.0
go: downloading github.com/gorilla/mux v1.8.0
go: downloading github.com/gorilla/websocket v1.5.0
go: downloading github.com/ipfs/go-metrics-interface v0.0.1
go: downloading github.com/ipni/index-provider v0.12.0
go: downloading github.com/libp2p/go-libp2p-gorpc v0.5.0
go: downloading github.com/libp2p/go-libp2p-record v0.2.0
go: downloading github.com/ipld/go-ipld-prime v0.20.0
go: downloading github.com/filecoin-project/dagstore v0.5.2
go: downloading github.com/raulk/go-watchdog v1.3.0
go: downloading github.com/Kubuxu/imtui v0.0.0-20210401140320-41663d68d0fa
go: downloading github.com/buger/goterm v1.0.3
go: downloading github.com/chzyer/readline v1.5.1
go: downloading github.com/dustin/go-humanize v1.0.1
go: downloading github.com/filecoin-project/go-cbor-util v0.0.1
go: downloading github.com/gdamore/tcell/v2 v2.2.0
go: downloading github.com/ipfs/go-cidutil v0.1.0
go: downloading github.com/ipld/go-ipld-selector-text-lite v0.0.1
go: downloading github.com/multiformats/go-multibase v0.2.0
go: downloading golang.org/x/term v0.9.0
go: downloading github.com/pkg/errors v0.9.1
go: downloading github.com/mattn/go-colorable v0.1.13
go: downloading github.com/golang/groupcache v0.0.0-20210331224755-41bb18bfe9da
go: downloading github.com/filecoin-project/kubo-api-client v0.0.1
go: downloading github.com/libp2p/go-msgio v0.3.0
go: downloading github.com/filecoin-project/go-crypto v0.0.1
go: downloading golang.org/x/crypto v0.10.0
go: downloading github.com/mattn/go-runewidth v0.0.10
go: downloading github.com/hashicorp/errwrap v1.1.0
go: downloading github.com/polydawn/refmt v0.89.0
go: downloading github.com/cpuguy83/go-md2man/v2 v2.0.2
go: downloading github.com/xrash/smetrics v0.0.0-20201216005158-039620a65673
go: downloading github.com/gogo/protobuf v1.3.2
go: downloading github.com/libp2p/go-buffer-pool v0.1.0
go: downloading github.com/go-logr/logr v1.2.4
go: downloading go.opentelemetry.io/otel/metric v1.16.0
go: downloading go.opentelemetry.io/otel/trace v1.16.0
go: downloading go.opentelemetry.io/otel/sdk/metric v0.39.0
go: downloading github.com/go-logr/stdr v1.2.2
go: downloading github.com/jbenet/goprocess v0.1.4
go: downloading github.com/prometheus/statsd_exporter v0.22.7
go: downloading github.com/beorn7/perks v1.0.1
go: downloading github.com/cespare/xxhash/v2 v2.2.0
go: downloading github.com/cespare/xxhash v1.1.0
go: downloading github.com/golang/protobuf v1.5.3
go: downloading github.com/prometheus/client_model v0.4.0
go: downloading github.com/prometheus/common v0.42.0
go: downloading github.com/prometheus/procfs v0.9.0
go: downloading google.golang.org/protobuf v1.30.0
go: downloading github.com/google/gopacket v1.1.19
go: downloading github.com/libp2p/go-libp2p-kbucket v0.6.1
go: downloading github.com/libp2p/go-netroute v0.2.1
go: downloading github.com/whyrusleeping/go-keyspace v0.0.0-20160322163242-5b898ac5add1
go: downloading gonum.org/v1/gonum v0.13.0
go: downloading github.com/filecoin-project/go-amt-ipld/v2 v2.1.0
go: downloading github.com/filecoin-project/go-hamt-ipld v0.1.5
go: downloading github.com/minio/sha256-simd v1.0.1
go: downloading github.com/filecoin-project/go-commp-utils/nonffi v0.0.0-20220905160352-62059082a837
go: downloading github.com/Gurpartap/async v0.0.0-20180927173644-4f7f499dd9ee
go: downloading github.com/filecoin-project/go-hamt-ipld/v2 v2.0.0
go: downloading github.com/filecoin-project/go-amt-ipld/v3 v3.1.0
go: downloading github.com/filecoin-project/go-hamt-ipld/v3 v3.1.0
go: downloading github.com/stretchr/testify v1.8.4
go: downloading github.com/elastic/gosigar v0.14.2
go: downloading go.uber.org/dig v1.17.0
go: downloading github.com/mr-tron/base58 v1.2.0
go: downloading github.com/multiformats/go-multicodec v0.9.0
go: downloading github.com/dgraph-io/ristretto v0.1.0
go: downloading github.com/dgryski/go-farm v0.0.0-20200201041132-a6ae2369ad13
go: downloading golang.org/x/net v0.10.0
go: downloading github.com/ipfs/go-merkledag v0.11.0
go: downloading github.com/ipld/go-codec-dagpb v1.6.0
go: downloading github.com/ipfs/go-ipfs-blockstore v1.3.0
go: downloading github.com/ipfs/go-ipfs-util v0.0.3
go: downloading go4.org v0.0.0-20230225012048-214862532bf5
go: downloading github.com/ipfs/go-ipld-legacy v0.2.1
go: downloading github.com/jpillora/backoff v1.0.0
go: downloading github.com/filecoin-project/go-statemachine v1.0.3
go: downloading github.com/multiformats/go-multiaddr-dns v0.3.1
go: downloading github.com/multiformats/go-multistream v0.4.1
go: downloading github.com/filecoin-project/go-ds-versioning v0.1.2
go: downloading github.com/hannahhoward/go-pubsub v0.0.0-20200423002714-8d62886cc36e
go: downloading github.com/whyrusleeping/ledger-filecoin-go v0.9.1-0.20201010031517-c3dcc1bddce4
go: downloading github.com/hashicorp/go-hclog v1.3.0
go: downloading github.com/hashicorp/raft v1.3.10
go: downloading github.com/hashicorp/raft-boltdb v0.0.0-20171010151810-6e5ba93211ea
go: downloading github.com/libp2p/go-libp2p-consensus v0.0.1
go: downloading github.com/libp2p/go-libp2p-raft v0.4.0
go: downloading golang.org/x/exp v0.0.0-20230321023759-10a507213a29
go: downloading github.com/filecoin-project/go-padreader v0.0.1
go: downloading github.com/zyedidia/generic v1.2.1
go: downloading github.com/libp2p/go-libp2p-routing-helpers v0.7.0
go: downloading github.com/libp2p/go-maddr-filter v0.1.0
go: downloading github.com/whyrusleeping/multiaddr-filter v0.0.0-20160516205228-e903e4adabd7
go: downloading github.com/containerd/cgroups v1.1.0
go: downloading github.com/elastic/go-sysinfo v1.7.0
go: downloading github.com/filecoin-project/go-storedcounter v0.1.0
go: downloading github.com/decred/dcrd/dcrec/secp256k1/v4 v4.2.0
go: downloading github.com/libp2p/go-flow-metrics v0.1.0
go: downloading github.com/bep/debounce v1.2.1
go: downloading github.com/filecoin-project/go-fil-commp-hashhash v0.1.0
go: downloading github.com/ipni/go-libipni v0.0.8
go: downloading github.com/drand/drand v1.4.9
go: downloading github.com/drand/kyber v1.1.15
go: downloading github.com/ipfs/go-ipfs-delay v0.0.1
go: downloading github.com/ipfs/bbloom v0.0.4
go: downloading github.com/acarl005/stripansi v0.0.0-20180116102854-5a71ef0e047d
go: downloading github.com/ipfs/go-unixfsnode v1.7.1
go: downloading github.com/benbjohnson/clock v1.3.5
go: downloading github.com/ugorji/go/codec v1.2.6
go: downloading github.com/ipfs/go-peertaskqueue v0.8.1
go: downloading github.com/gdamore/encoding v1.0.0
go: downloading github.com/lucasb-eyer/go-colorful v1.0.3
go: downloading golang.org/x/text v0.10.0
go: downloading github.com/russross/blackfriday/v2 v2.1.0
go: downloading github.com/multiformats/go-base36 v0.2.0
go: downloading github.com/go-kit/log v0.2.1
go: downloading gopkg.in/yaml.v2 v2.4.0
go: downloading github.com/ipld/go-ipld-adl-hamt v0.0.0-20220616142416-9004dbd839e0
go: downloading github.com/opentracing/opentracing-go v1.2.0
go: downloading github.com/matttproud/golang_protobuf_extensions v1.0.4
go: downloading github.com/ipfs/go-ipfs-cmds v0.9.0
go: downloading github.com/rivo/uniseg v0.1.0
go: downloading github.com/ipsn/go-secp256k1 v0.0.0-20180726113642-9d62b9f0bc52
go: downloading github.com/golang/snappy v0.0.4
go: downloading github.com/klauspost/compress v1.16.5
go: downloading github.com/golang/glog v1.1.0
go: downloading github.com/spaolacci/murmur3 v1.1.0
go: downloading github.com/klauspost/cpuid/v2 v2.2.5
go: downloading github.com/ipfs/go-ipfs-ds-help v1.1.0
go: downloading github.com/libp2p/go-cidranger v1.1.0
go: downloading github.com/libp2p/go-libp2p-asn-util v0.3.0
go: downloading github.com/boltdb/bolt v1.3.1
go: downloading github.com/hashicorp/go-msgpack v0.5.5
go: downloading github.com/petar/GoLLRB v0.0.0-20210522233825-ae3b015fd3e9
go: downloading github.com/whyrusleeping/cbor v0.0.0-20171005072247-63513f603b11
go: downloading github.com/ipfs/go-blockservice v0.5.1
go: downloading github.com/crackcomm/go-gitignore v0.0.0-20170627025303-887ab5e44cc3
go: downloading github.com/pbnjay/memory v0.0.0-20210728143218-7b4eea64cf58
go: downloading github.com/quic-go/quic-go v0.33.1
go: downloading github.com/elastic/go-elasticsearch/v7 v7.14.0
go: downloading github.com/miekg/dns v1.1.54
go: downloading github.com/multiformats/go-multiaddr-fmt v0.1.0
go: downloading github.com/libp2p/go-yamux/v4 v4.0.0
go: downloading github.com/flynn/noise v1.0.0
go: downloading github.com/zondax/ledger-go v0.12.1
go: downloading github.com/libp2p/go-libp2p-gostream v0.6.0
go: downloading github.com/armon/go-metrics v0.3.9
go: downloading github.com/coreos/go-systemd/v22 v22.5.0
go: downloading github.com/godbus/dbus/v5 v5.1.0
go: downloading github.com/opencontainers/runtime-spec v1.0.2
go: downloading github.com/cilium/ebpf v0.9.1
go: downloading github.com/sirupsen/logrus v1.9.0
go: downloading github.com/ipfs/go-ipfs-pq v0.0.3
go: downloading github.com/go-logfmt/logfmt v0.5.1
go: downloading golang.org/x/time v0.0.0-20220722155302-e5dcc9cfc0b9
go: downloading github.com/davecgh/go-spew v1.1.1
go: downloading github.com/pmezard/go-difflib v1.0.0
go: downloading gopkg.in/yaml.v3 v3.0.1
go: downloading github.com/nikkolasg/hexjson v0.1.0
go: downloading github.com/whyrusleeping/chunker v0.0.0-20181014151217-fe64bd25879f
go: downloading github.com/alecthomas/units v0.0.0-20211218093645-b94a6e3cc137
go: downloading github.com/libp2p/go-nat v0.1.0
go: downloading github.com/quic-go/webtransport-go v0.5.3
go: downloading github.com/twmb/murmur3 v1.1.6
go: downloading github.com/jbenet/go-temp-err-catcher v0.1.0
go: downloading github.com/libp2p/go-reuseport v0.2.0
go: downloading github.com/marten-seemann/tcp v0.0.0-20210406111302-dfbc87cc63fd
go: downloading github.com/mikioh/tcpinfo v0.0.0-20190314235526-30a79bb1804b
go: downloading nhooyr.io/websocket v1.8.7
go: downloading github.com/cskr/pubsub v1.0.2
go: downloading lukechampine.com/blake3 v1.2.1
go: downloading github.com/rs/cors v1.7.0
go: downloading howett.net/plist v0.0.0-20181124034731-591f970eefbb
go: downloading github.com/joeshaw/multierror v0.0.0-20140124173710-69b34d4ec901
go: downloading github.com/ipfs/go-ipfs-exchange-interface v0.2.0
go: downloading github.com/ipfs/go-verifcid v0.0.2
go: downloading github.com/ipfs/go-bitfield v1.1.0
go: downloading github.com/drand/kyber-bls12381 v0.2.3
go: downloading google.golang.org/grpc v1.55.0
go: downloading github.com/zondax/hid v0.9.1
go: downloading github.com/davidlazar/go-crypto v0.0.0-20200604182044-b73af7476f6c
go: downloading github.com/hashicorp/go-immutable-radix v1.3.1
go: downloading github.com/huin/goupnp v1.2.0
go: downloading github.com/jackpal/go-nat-pmp v1.0.2
go: downloading github.com/koron/go-ssdp v0.0.4
go: downloading github.com/mikioh/tcpopt v0.0.0-20190314235656-172688c1accc
go: downloading github.com/quic-go/qpack v0.4.0
go: downloading github.com/francoispqt/gojay v1.2.13
go: downloading github.com/quic-go/qtls-go1-19 v0.3.3
go: downloading google.golang.org/genproto v0.0.0-20230306155012-7f2fa6fef1f4
go: downloading github.com/kilic/bls12-381 v0.1.0
# runtime/internal/sys
/usr/local/go/src/runtime/internal/sys/consts.go:16:7: DefaultPhysPageSize redeclared in this block
        /usr/local/go/src/runtime/internal/sys/arch_amd64.go:10:2: other declaration of DefaultPhysPageSize
/usr/local/go/src/runtime/internal/sys/consts.go:20:7: PCQuantum redeclared in this block
        /usr/local/go/src/runtime/internal/sys/arch_amd64.go:11:2: other declaration of PCQuantum
/usr/local/go/src/runtime/internal/sys/consts.go:23:7: Int64Align redeclared in this block
        /usr/local/go/src/runtime/internal/sys/arch_amd64.go:12:2: other declaration of Int64Align
/usr/local/go/src/runtime/internal/sys/consts.go:30:7: MinFrameSize redeclared in this block
        /usr/local/go/src/runtime/internal/sys/arch_amd64.go:13:2: other declaration of MinFrameSize
/usr/local/go/src/runtime/internal/sys/stubs.go:16:7: StackGuardMultiplier redeclared in this block
        /usr/local/go/src/runtime/internal/sys/consts.go:13:7: other declaration of StackGuardMultiplier
# math
/usr/local/go/src/math/acosh.go:43:6: Acosh defined in both Go and assembly
/usr/local/go/src/math/asin.go:20:6: Asin defined in both Go and assembly
/usr/local/go/src/math/asin.go:58:6: Acos defined in both Go and assembly
/usr/local/go/src/math/asinh.go:40:6: Asinh defined in both Go and assembly
/usr/local/go/src/math/atan.go:96:6: Atan defined in both Go and assembly
/usr/local/go/src/math/atan2.go:30:6: Atan2 defined in both Go and assembly
/usr/local/go/src/math/atanh.go:48:6: Atanh defined in both Go and assembly
/usr/local/go/src/math/cbrt.go:26:6: Cbrt defined in both Go and assembly
/usr/local/go/src/math/erf.go:189:6: Erf defined in both Go and assembly
/usr/local/go/src/math/erf.go:274:6: Erfc defined in both Go and assembly
/usr/local/go/src/math/erf.go:274:6: too many errors
make: *** [Makefile:89: lotus] Error 1`

**Comment by TippyFlitsUK** [comment]
Did you fully uninstall the old GO before installing the new version?

**Reply by PNetrox** [reply]
Yep, i forgot about that, but now i have new issue:
`/home/filecoin/lotus/extern/filecoin-ffi
+ ls ./filcrypto.h ./filcrypto.pc ./install-filcrypto ./libfilcrypto.a
./filcrypto.h  ./filcrypto.pc  ./install-filcrypto  ./libfilcrypto.a
+ [[ ! -f ./filcrypto.h ]]
+ [[ ! -f ./libfilcrypto.a ]]
+ [[ ! -f ./filcrypto.pc ]]
+ echo '[install-filcrypto/main] successfully built and installed libfilcrypto from source'
[install-filcrypto/main] successfully built and installed libfilcrypto from source
+ exit
make[1]: Leaving directory '/home/filecoin/lotus/extern/filecoin-ffi'
rm -f lotus
go build  -ldflags="-X=github.com/filecoin-project/lotus/build.CurrentCommit=+git.7bb1f98ac" -tags=calibnet -o lotus ./cmd/lotus
# internal/abi
/usr/local/go/src/internal/abi/funcpc.go:22:6: FuncPCABI0 redeclared in this block
        /usr/local/go/src/internal/abi/abi.go:117:6: other declaration of FuncPCABI0
/usr/local/go/src/internal/abi/funcpc.go:31:6: FuncPCABIInternal redeclared in this block
        /usr/local/go/src/internal/abi/abi.go:126:6: other declaration of FuncPCABIInternal
# hash/maphash
/usr/local/go/src/hash/maphash/maphash_runtime.go:14:6: runtime_fastrand64 redeclared in this block
        /usr/local/go/src/hash/maphash/maphash.go:268:6: other declaration of runtime_fastrand64
/usr/local/go/src/hash/maphash/maphash_runtime.go:18:6: runtime_memhash redeclared in this block
        /usr/local/go/src/hash/maphash/maphash.go:287:6: other declaration of runtime_memhash
/usr/local/go/src/hash/maphash/maphash_runtime.go:20:6: rthash redeclared in this block
        /usr/local/go/src/hash/maphash/maphash.go:270:6: other declaration of rthash
/usr/local/go/src/hash/maphash/maphash_runtime.go:37:29: undefined: unsafe.StringData
make: *** [Makefile:89: lotus] Error 1`

**Reply by TippyFlitsUK** [reply]
How did you delete the old GO version?

**Reply by PNetrox** [reply]
as root

`rm -rf /usr/local/go`

**Reply by PNetrox** [reply]
Now im trying it again and again, and im think now im close:

`go: downloading github.com/quic-go/qtls-go1-20 v0.2.3
# github.com/quic-go/quic-go/internal/qtls
../go/pkg/mod/github.com/quic-go/quic-go@v0.33.1/internal/qtls/go121.go:5:13: cannot use "The version of quic-go you're using can't be built on Go 1.21 yet. For more details, please see https://github.com/quic-go/quic-go/wiki/quic-go-and-Go-versions." (untyped string constant "The version of quic-go you're using can't be built on Go 1.21 yet. F...) as int value in variable declaration
make: *** [Makefile:89: lotus] Error 1`

**Reply by PNetrox** [reply]
Whats the best go version for lotus?

**Reply by TippyFlitsUK** [reply, marked as answer]
Use 1.20.7

**Reply by PNetrox** [reply]
Now it works, thanks

**Comment by PNetrox** [comment]
> Use 1.20.7
